### ISSUE TITLE
Fatal Python error: ffi.from_handle() detected that the address passed points to garbage

### ISSUE THREAD

**Issue by garlick** [flux-framework/flux-core#629]
Python tests are hanging on TOSS 3 system:
```
PASS: ../t/t9990-python-tests.t 1 test.barrier.TestBarrier.test_eight
PASS: ../t/t9990-python-tests.t 2 test.barrier.TestBarrier.test_single
PASS: ../t/t9990-python-tests.t 3 test.event.TestEvent.test_full_event
PASS: ../t/t9990-python-tests.t 4 test.event.TestEvent.test_t1_0_sub
PASS: ../t/t9990-python-tests.t 5 test.event.TestEvent.test_t1_1_unsub
PASS: ../t/t9990-python-tests.t 6 test.handle.TestHandle.test_create_handle
PASS: ../t/t9990-python-tests.t 7 test.handle.TestHandle.test_log
PASS: ../t/t9990-python-tests.t 8 test.handle.TestHandle.test_rpc_ping
PASS: ../t/t9990-python-tests.t 9 test.json_t.TestJsonCInterface.test_create_empty
PASS: ../t/t9990-python-tests.t 10 test.json_t.TestJsonCInterface.test_with_data
PASS: ../t/t9990-python-tests.t 11 test.kvs.TestKVS.test_bad_init
PASS: ../t/t9990-python-tests.t 12 test.kvs.TestKVS.test_exists_dir
PASS: ../t/t9990-python-tests.t 13 test.kvs.TestKVS.test_exists_false
PASS: ../t/t9990-python-tests.t 14 test.kvs.TestKVS.test_exists_true
PASS: ../t/t9990-python-tests.t 15 test.kvs.TestKVS.test_fill
PASS: ../t/t9990-python-tests.t 16 test.kvs.TestKVS.test_iterator
PASS: ../t/t9990-python-tests.t 17 test.kvs.TestKVS.test_key_at
PASS: ../t/t9990-python-tests.t 18 test.kvs.TestKVS.test_kvs_dir_open
PASS: ../t/t9990-python-tests.t 19 test.kvs.TestKVS.test_read_non_existent
PASS: ../t/t9990-python-tests.t 20 test.kvs.TestKVS.test_read_non_existent_basedir
PASS: ../t/t9990-python-tests.t 21 test.kvs.TestKVS.test_remove
PASS: ../t/t9990-python-tests.t 22 test.kvs.TestKVS.test_set_deep
PASS: ../t/t9990-python-tests.t 23 test.kvs.TestKVS.test_set_dict
PASS: ../t/t9990-python-tests.t 24 test.kvs.TestKVS.test_set_float
PASS: ../t/t9990-python-tests.t 25 test.kvs.TestKVS.test_set_int
PASS: ../t/t9990-python-tests.t 26 test.kvs.TestKVS.test_set_list
PASS: ../t/t9990-python-tests.t 27 test.kvs.TestKVS.test_set_none
PASS: ../t/t9990-python-tests.t 28 test.kvs.TestKVS.test_set_string
PASS: ../t/t9990-python-tests.t 29 test.kvs.TestKVS.test_walk
PASS: ../t/t9990-python-tests.t 30 test.kvs.TestKVS.test_walk_with_no_handle
PASS: ../t/t9990-python-tests.t 31 test.request.TestRequestMethods.test_no_topic_invalid
PASS: ../t/t9990-python-tests.t 32 test.request.TestRequestMethods.test_null_payload
PASS: ../t/t9990-python-tests.t 33 test.watchers.TestTimer.test_s1_0_timer_add
PASS: ../t/t9990-python-tests.t 34 test.watchers.TestTimer.test_s1_1_timer_remove
PASS: ../t/t9990-python-tests.t 35 test.watchers.TestTimer.test_timer_add_negative
```
after waiting many minutes, I hit ctrl-C
```
^Cmake[3]: *** Deleting file `../t/t9990-python-tests.log'tap-driver.sh: fatal: I/O or internal error
make[3]: *** [../t/t9990-python-tests.log] Error 1
make[2]: *** [check-TESTS] Interrupt
make[1]: *** [check-am] Interrupt
make: *** [check-recursive] Interrupt
```

This seems  to reproduce (three for three tries at this point).   This is on current master `0.2.0-229-gd1ddd8d`

On my desktop the next test would be
```
PASS: t9990-python-tests.t 34 test.watchers.TestTimer.test_timer_with_reactor
```
(Note tests appear to run in a different order there - barrier tests show up last not first)

**Comment by garlick** [comment]
Running the python test driver manually:
```
TAP version 13
'#Running doctests and unit tests in: /g/g0/garlick/flux-core/src/bindings/python/test/barrier.py
ok 1 test.barrier.TestBarrier.test_eight
ok 2 test.barrier.TestBarrier.test_single
#Running doctests and unit tests in: /g/g0/garlick/flux-core/src/bindings/python/test/event.py
ok 3 test.event.TestEvent.test_full_event
# {"test": "yay!"}
ok 4 test.event.TestEvent.test_t1_0_sub
ok 5 test.event.TestEvent.test_t1_1_unsub
#Running doctests and unit tests in: /g/g0/garlick/flux-core/src/bindings/python/test/handle.py
ok 6 test.handle.TestHandle.test_create_handle
ok 7 test.handle.TestHandle.test_log
ok 8 test.handle.TestHandle.test_rpc_ping
#Running doctests and unit tests in: /g/g0/garlick/flux-core/src/bindings/python/test/json_t.py
ok 9 test.json_t.TestJsonCInterface.test_create_empty
ok 10 test.json_t.TestJsonCInterface.test_with_data
#Running doctests and unit tests in: /g/g0/garlick/flux-core/src/bindings/python/test/kvs.py
ok 11 test.kvs.TestKVS.test_bad_init
ok 12 test.kvs.TestKVS.test_exists_dir
ok 13 test.kvs.TestKVS.test_exists_false
ok 14 test.kvs.TestKVS.test_exists_true
ok 15 test.kvs.TestKVS.test_fill
ok 16 test.kvs.TestKVS.test_iterator
# ['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14']
# passed 1
# passed 2
# passed 3
# passed 4
# passed 5
# passed 6
# passed 7
# passed 8
# passed 9
# passed 10
# passed 11
# passed 12
# passed 13
# passed 14
ok 17 test.kvs.TestKVS.test_key_at
ok 18 test.kvs.TestKVS.test_kvs_dir_open
ok 19 test.kvs.TestKVS.test_read_non_existent
ok 20 test.kvs.TestKVS.test_read_non_existent_basedir
ok 21 test.kvs.TestKVS.test_remove
ok 22 test.kvs.TestKVS.test_set_deep
ok 23 test.kvs.TestKVS.test_set_dict
ok 24 test.kvs.TestKVS.test_set_float
ok 25 test.kvs.TestKVS.test_set_int
ok 26 test.kvs.TestKVS.test_set_list
ok 27 test.kvs.TestKVS.test_set_none
ok 28 test.kvs.TestKVS.test_set_string
ok 29 test.kvs.TestKVS.test_walk
# ('', <generator object directories at 0xd36320>, <generator object files at 0xd36370>)
ok 30 test.kvs.TestKVS.test_walk_with_no_handle
#Running doctests and unit tests in: /g/g0/garlick/flux-core/src/bindings/python/test/request.py
ok 31 test.request.TestRequestMethods.test_no_topic_invalid
ok 32 test.request.TestRequestMethods.test_null_payload
#Running doctests and unit tests in: /g/g0/garlick/flux-core/src/bindings/python/test/watchers.py
ok 33 test.watchers.TestTimer.test_s1_0_timer_add
ok 34 test.watchers.TestTimer.test_s1_1_timer_remove
ok 35 test.watchers.TestTimer.test_timer_add_negative
Fatal Python error: ffi.from_handle() detected that the address passed points to garbage. If it is really the result of ffi.new_handle(), then the Python object has already been garbage collected
./t9990-python-tests.t: line 8:  3229 Aborted                 (core dumped) ../src/bindings/python/test_commands/test_runner.t
```
A core file was left - not sure if this bt is useful
```
Using host libthread_db library "/lib64/libthread_db.so.1".
Core was generated by `python ../src/bindings/python/test_commands/test_runner.t'.
Program terminated with signal 6, Aborted.
#0  0x00002aaaab9ef5f7 in __GI_raise (sig=sig@entry=6)
    at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
56	  return INLINE_SYSCALL (tgkill, 3, pid, selftid, sig);
Missing separate debuginfos, use: debuginfo-install python-2.7.5-34.el7.x86_64
(gdb) bt
#0  0x00002aaaab9ef5f7 in __GI_raise (sig=sig@entry=6)
    at ../nptl/sysdeps/unix/sysv/linux/raise.c:56
#1  0x00002aaaab9f0ce8 in __GI_abort () at abort.c:90
#2  0x00002aaaaadcbaee in Py_FatalError () from /lib64/libpython2.7.so.1.0
#3  0x00002aaab05a5a37 in b_from_handle ()
   from /usr/lib64/python2.7/site-packages/_cffi_backend.so
#4  0x00002aaaaadafc5c in PyEval_EvalFrameEx () from /lib64/libpython2.7.so.1.0
#5  0x00002aaaaadb11ed in PyEval_EvalCodeEx () from /lib64/libpython2.7.so.1.0
#6  0x00002aaaaad3e0c8 in function_call () from /lib64/libpython2.7.so.1.0
#7  0x00002aaaaad190c3 in PyObject_Call () from /lib64/libpython2.7.so.1.0
#8  0x00002aaab05b0a5c in general_invoke_callback ()
   from /usr/lib64/python2.7/site-packages/_cffi_backend.so
#9  0x00002aaab05b0d31 in invoke_callback ()
   from /usr/lib64/python2.7/site-packages/_cffi_backend.so
#10 0x00002aaab07ccb9b in ffi_closure_unix64_inner () from /lib64/libffi.so.6
#11 0x00002aaab07ccf14 in ffi_closure_unix64 () from /lib64/libffi.so.6
#12 0x00002aaab0cbcf95 in ev_invoke_pending (loop=0xdea2f0) at ev.c:3088
#13 0x00002aaab0cc00f7 in ev_run (loop=0xdea2f0, flags=0) at ev.c:3488
#14 0x00002aaab0ca194d in flux_reactor_run (r=0xb87750, flags=flags@entry=0)
    at reactor.c:129
#15 0x00002aaab09f9c96 in _cffi_f_flux_reactor_run (self=<optimized out>, 
    args=<optimized out>) at _core.c:9372
#16 0x00002aaaaadb0702 in PyEval_EvalFrameEx () from /lib64/libpython2.7.so.1.0
#17 0x00002aaaaadb11ed in PyEval_EvalCodeEx () from /lib64/libpython2.7.so.1.0
#18 0x00002aaaaad3e0c8 in function_call () from /lib64/libpython2.7.so.1.0
#19 0x00002aaaaad190c3 in PyObject_Call () from /lib64/libpython2.7.so.1.0
#20 0x00002aaaaad280b5 in instancemethod_call ()
   from /lib64/libpython2.7.so.1.0
#21 0x00002aaaaad190c3 in PyObject_Call () from /lib64/libpython2.7.so.1.0
#22 0x00002aaaaad70527 in slot_tp_call () from /lib64/libpython2.7.so.1.0
#23 0x00002aaaaad190c3 in PyObject_Call () from /lib64/libpython2.7.so.1.0
#24 0x00002aaaaadad38c in PyEval_EvalFrameEx () from /lib64/libpython2.7.so.1.0
#25 0x00002aaaaadb11ed in PyEval_EvalCodeEx () from /lib64/libpython2.7.so.1.0
#26 0x00002aaaaadaf89f in PyEval_EvalFrameEx () from /lib64/libpython2.7.so.1.0
#27 0x00002aaaaadb11ed in PyEval_EvalCodeEx () from /lib64/libpython2.7.so.1.0
#28 0x00002aaaaad3e1bd in function_call () from /lib64/libpython2.7.so.1.0
#29 0x00002aaaaad190c3 in PyObject_Call () from /lib64/libpython2.7.so.1.0
#30 0x00002aaaaadac427 in PyEval_EvalFrameEx () from /lib64/libpython2.7.so.1.0
#31 0x00002aaaaadb11ed in PyEval_EvalCodeEx () from /lib64/libpython2.7.so.1.0
#32 0x00002aaaaad3e0c8 in function_call () from /lib64/libpython2.7.so.1.0
#33 0x00002aaaaad190c3 in PyObject_Call () from /lib64/libpython2.7.so.1.0
#34 0x00002aaaaad280b5 in instancemethod_call ()
   from /lib64/libpython2.7.so.1.0
#35 0x00002aaaaad190c3 in PyObject_Call () from /lib64/libpython2.7.so.1.0
#36 0x00002aaaaad70527 in slot_tp_call () from /lib64/libpython2.7.so.1.0
#37 0x00002aaaaad190c3 in PyObject_Call () from /lib64/libpython2.7.so.1.0
#38 0x00002aaaaadad38c in PyEval_EvalFrameEx () from /lib64/libpython2.7.so.1.0
#39 0x00002aaaaadb11ed in PyEval_EvalCodeEx () from /lib64/libpython2.7.so.1.0
#40 0x00002aaaaad3e1bd in function_call () from /lib64/libpython2.7.so.1.0
#41 0x00002aaaaad190c3 in PyObject_Call () from /lib64/libpython2.7.so.1.0
#42 0x00002aaaaadac427 in PyEval_EvalFrameEx () from /lib64/libpython2.7.so.1.0
#43 0x00002aaaaadb11ed in PyEval_EvalCodeEx () from /lib64/libpython2.7.so.1.0
#44 0x00002aaaaad3e0c8 in function_call () from /lib64/libpython2.7.so.1.0
#45 0x00002aaaaad190c3 in PyObject_Call () from /lib64/libpython2.7.so.1.0
#46 0x00002aaaaad280b5 in instancemethod_call ()
   from /lib64/libpython2.7.so.1.0
#47 0x00002aaaaad190c3 in PyObject_Call () from /lib64/libpython2.7.so.1.0
#48 0x00002aaaaad70527 in slot_tp_call () from /lib64/libpython2.7.so.1.0
#49 0x00002aaaaad190c3 in PyObject_Call () from /lib64/libpython2.7.so.1.0
#50 0x00002aaaaadad38c in PyEval_EvalFrameEx () from /lib64/libpython2.7.so.1.0
#51 0x00002aaaaadb11ed in PyEval_EvalCodeEx () from /lib64/libpython2.7.so.1.0
#52 0x00002aaaaad3e1bd in function_call () from /lib64/libpython2.7.so.1.0
#53 0x00002aaaaad190c3 in PyObject_Call () from /lib64/libpython2.7.so.1.0
#54 0x00002aaaaadac427 in PyEval_EvalFrameEx () from /lib64/libpython2.7.so.1.0
#55 0x00002aaaaadb11ed in PyEval_EvalCodeEx () from /lib64/libpython2.7.so.1.0
#56 0x00002aaaaad3e0c8 in function_call () from /lib64/libpython2.7.so.1.0
#57 0x00002aaaaad190c3 in PyObject_Call () from /lib64/libpython2.7.so.1.0
#58 0x00002aaaaad280b5 in instancemethod_call ()
   from /lib64/libpython2.7.so.1.0
#59 0x00002aaaaad190c3 in PyObject_Call () from /lib64/libpython2.7.so.1.0
#60 0x00002aaaaad70527 in slot_tp_call () from /lib64/libpython2.7.so.1.0
#61 0x00002aaaaad190c3 in PyObject_Call () from /lib64/libpython2.7.so.1.0
#62 0x00002aaaaadad38c in PyEval_EvalFrameEx () from /lib64/libpython2.7.so.1.0
#63 0x00002aaaaadaf990 in PyEval_EvalFrameEx () from /lib64/libpython2.7.so.1.0
#64 0x00002aaaaadb11ed in PyEval_EvalCodeEx () from /lib64/libpython2.7.so.1.0
#65 0x00002aaaaadaf89f in PyEval_EvalFrameEx () from /lib64/libpython2.7.so.1.0
#66 0x00002aaaaadb11ed in PyEval_EvalCodeEx () from /lib64/libpython2.7.so.1.0
#67 0x00002aaaaadb12f2 in PyEval_EvalCode () from /lib64/libpython2.7.so.1.0
#68 0x00002aaaaadca72f in run_mod () from /lib64/libpython2.7.so.1.0
#69 0x00002aaaaadcb8ee in PyRun_FileExFlags () from /lib64/libpython2.7.so.1.0
#70 0x00002aaaaadccb79 in PyRun_SimpleFileExFlags ()
   from /lib64/libpython2.7.so.1.0
#71 0x00002aaaaadddcaf in Py_Main () from /lib64/libpython2.7.so.1.0
#72 0x00002aaaab9dbb15 in __libc_start_main (main=0x4006f0 <main>, argc=2, 
    ubp_av=0x7fffffffdba8, init=<optimized out>, fini=<optimized out>, 
    rtld_fini=<optimized out>, stack_end=0x7fffffffdb98) at libc-start.c:274
#73 0x0000000000400721 in _start ()
(gdb) 
```

**Comment by garlick** [comment]
These tests were run on "upgrade", with
```
python-cffi-1.5.0-2.ch6.x86_64
python-2.7.5-34.el7.x86_64
python-devel-2.7.5-34.el7.x86_64
```

**Comment by garlick** [comment]
It would be good to have this solved for the flux-core 0.3.0 release (just tagged it with that milestone).  I think we may need some of @trws's expertise here.

**Comment by trws** [comment]
Pretty sure I've got this worked out.  It looks like I accidentally removed a back reference I was using to keep the reference to watchers from getting garbage collected when I updated the watcher interfaces a while back.  For some reason that didn't matter for older versions of Python, but does for the version in TOSS3.  Either way there should have been references there the whole time, so the patch just adds a back reference in each watcher constructor.  There is a small chance this could result in a dangling object, but the Python cycle catcher should always get it since the cycle is only two objects long.
